core/execution/future/market.go

```diff
@@ -997,6 +999,86 @@ func (m *Market) parkAllPeggedOrders(ctx context.Context) []*types.Order {
 	return parked
 }
 
+func (m *Market) uncrossOrderAtAuctionEnd(ctx context.Context) {
+	if !m.as.InAuction() || m.as.IsOpeningAuction() {
+		return
+	}
+	// Change market type to continuous trading
+	uncrossedOrders, _, err := m.matching.LeaveAuction(m.timeService.GetTimeNow())
+	if err != nil {
+		m.log.Error("Error leaving auction", logging.Error(err))
+	}
+
+	// Process each confirmation & apply fee calculations to each trade
+	evts := make([]events.Event, 0, len(uncrossedOrders))
+	for _, uncrossedOrder := range uncrossedOrders {
+		// handle fees first
+		err := m.applyFees(ctx, uncrossedOrder.Order, uncrossedOrder.Trades)
+		if err != nil {
+			// @TODO this ought to be an event
+			m.log.Error("Unable to apply fees to order",
+				logging.String("OrderID", uncrossedOrder.Order.ID))
+		}
+
+		// then do the confirmation
+		m.handleConfirmation(ctx, uncrossedOrder)
+
+		if uncrossedOrder.Order.Remaining == 0 {
+			uncrossedOrder.Order.Status = types.OrderStatusFilled
+		}
+		evts = append(evts, events.NewOrderEvent(ctx, uncrossedOrder.Order))
+	}
+
+	// send order events in a single batch, it's more efficient
+	m.broker.SendBatch(evts)
+}
+
+func (m *Market) UpdateMarketState(ctx context.Context, changes *types.MarketStateUpdateConfiguration) error {
+	if changes.UpdateType == types.MarketStateUpdateTypeTerminate {
+		m.uncrossOrderAtAuctionEnd(ctx)
+		// terminate and settle
+		m.tradingTerminatedWithFinalState(ctx, types.MarketStateClosed, num.UintZero().Mul(changes.SettlementPrice, m.priceFactor))
+	} else if changes.UpdateType == types.MarketStateUpdateTypeSuspend {
+		m.mkt.State = types.MarketStateSuspendedViaGovernance
+		m.mkt.TradingMode = types.MarketTradingModeSuspendedViaGovernance
+		if m.as.InAuction() {
+			m.as.ExtendAuctionSuspension(types.AuctionDuration{Duration: int64(m.minDuration)})
+			evt := m.as.AuctionExtended(ctx, m.timeService.GetTimeNow())
+			if evt != nil {
+				m.broker.Send(evt)
+			}
+		} else {
+			m.as.StartGovernanceSuspensionAuction(m.timeService.GetTimeNow())
+			m.enterAuction(ctx)
+			m.broker.Send(events.NewMarketUpdatedEvent(ctx, *m.mkt))
+		}
+	} else if changes.UpdateType == types.MarketStateUpdateTypeResume && m.mkt.State == types.MarketStateSuspendedViaGovernance {
+		if m.as.GetState().Trigger == types.AuctionTriggerGovernanceSuspension {
+			m.as.EndGovernanceSuspensionAuction()
```

Comment:
Just because the auction was triggered with a governance vote doesn't imply that we can always leave it with a resume vote.

Some other triggers might be active now (the volume in the auction could be such that the liquidity trigger is now active or the uncrossing price could be triggering the price monitoring auction).

This should go through the regular `checkAuction` flow (but it'd need to get modified first). I think changing the market state to `Suspended` (from `GovernanceSuspended`) upon enactment of the resume proposal and then just going through `checkAuction` might just work.

Reply:
the idea is that if the trigger was suspension and there is not extension, we should just EndGovernanceSuspensionAuction and leave the auction (line 1056 is missing the check for the extension trigger, but imagine it's there)

Reply:
But the principle we want to follow is that we never leave an auction if we'd immediately re-enter another auction type, we just extend the original auction instead. 

So the trigger could be a suspension, but we there might be an extension when we try to resume the market.

Say:
- market is in continuous trading mode, price monitoring bounds are [95,105],
- market gets suspended via governance,
- orders get put in whilst market is in (governance) auction mode,
- orders are such that the uncrossing price is 106,
- in that case the original (governance suspension) auction should get extended by the price monitoring auction.

Reply:
again, no. I'm saying we enter this code only if the trigger is governance and there is no extension. 

```
if m.as.GetState().Trigger == types.AuctionTriggerGovernanceSuspension && m.as.GetState().Extension == types.AuctionTriggerUnspecified {
			m.as.EndGovernanceSuspensionAuction()
			m.leaveAuction(ctx, m.timeService.GetTimeNow())
		} 
```

Reply:
Aren't you referring to local changes by any chance? Can't find the snipped you've pasted above anywhere.

Reply:
I'm referring to the concept and I did say:  **(line 1056 is missing the check for the extension trigger, but imagine it's there)** - so now you don't need to imagine. I didn't want to push the changes before you agree that it makes sense.

Reply:
But when will that extension trigger get a chance to get set if:
https://github.com/vegaprotocol/vega/blob/81400c8187318b7db7e688dcdca4c3ad49481c46/core/execution/future/auction.go#L34-L36
?

Reply:
Yes, perfect.

Sorry, I somehow manage to convince myself there's a return here:
https://github.com/vegaprotocol/vega/blob/81400c8187318b7db7e688dcdca4c3ad49481c46/core/execution/future/auction.go#L34-L36